[MRG] MNT: Minor tweaks to enet_coordinate_descent

jakirkham opened the pull request:
Makes some small tweaks to `enet_coordinate_descent` regarding conditions to avoid unnecessary computation and to simplify the code for readability.

jakirkham commented:
Thoughts on this?

agramfort commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -232,23 +232,20 @@ def enet_coordinate_descent(floating[::1] w,
                     w[ii] = (fsign(tmp) * fmax(fabs(tmp) - alpha, 0)
                              / (norm_cols_X[ii] + beta))
 
-                if w[ii] != 0.0:
```
it's very common that w[ii] will be zero after soft thresholding so I suspect this slows up things.

jakirkham replied:
Also not following this either unfortunately. Given the condition in the `if`, the `else` case can only make `w[ii] != 0`. So the concern raised shouldn't actually occur. Feel free to clarify if I'm missing something though.

jakirkham replied:
Exactly. This is being run in the `else`, which does not set `w[ii]` to `0.0`.

amueller replied:
This is not in the "else", it's after it. Also, the "else" contains ``fmax(fabs(tmp) - alpha, 0)`` which is the soft-thresholding operation, so it's "likely" to result in 0.

agramfort commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -232,23 +232,20 @@ def enet_coordinate_descent(floating[::1] w,
                 tmp = dot(n_samples, &X[0, ii], 1, &R[0], 1)
 
-                if positive and tmp < 0:
+                if (positive or -alpha <= tmp) and (tmp <= alpha):
```
to keep the idea of this check I would just write

`if positive and tmp < alpha:`

otherwise it's quite redundant with the else code block.

jakirkham replied:
Not sure I follow. This condition catches all cases where `w[ii]` would be zero to avoid the computation in the `else` block.

amueller replied:
I would benchmark all of these? Otherwise no way to know if any of them are faster than the current code, right?

jakirkham commented:
Thanks for the feedback. Tried to clarify some things inline. Please let me know if that helped or if we are still missing something.

jakirkham committed:
Shortcut w[ii] computation that will be 0

While its true that having `tmp` less than 0, will result in `w[ii]`
being `0`, this is also true if `tmp` is less than `alpha`. So change
the check to look for `tmp` less than or equal to `alpha`.

jakirkham committed:
Break up positivity check when determining `w[ii]`

If `tmp` is positive semidefinite and less than `alpha`, it will effectively set
`w[ii]` to `0` through computation regardless of whether positivity is
enforced or not. Hence we check for this case alone.

If positivity is enforced, then `w[ii]` is set to `0` if `tmp` is
negative as well. So the first check is sufficient.

If positivity is not enforced, then `tmp` could be larger in magnitude
than `alpha`, which would be enough to make `w[ii]` not `0`. Hence we
need to check for this case as well.

In other words, if `tmp` is bounded by `-alpha` and `alpha`, `w[ii]`
will be `0` (regardless of whether positivity is enforced). If
positivity is enforced, we have the additional constraint that `tmp`
cannot be less than `0`, which means `tmp` must only be greater than
`alpha`.

jakirkham committed:
Merge `axpy` computation with `w[ii]` computation

As the first case of this `if` sets `w[ii]` to `0`, we already know it
doesn't apply to the `axpy` call, so we can merge this into the `else`
easily. Further if the first condition doesn't apply, we know `w[ii]` is
nonzero. So there is no need to check for it again within the `else`.
Hence we can cleanly merge the `axpy` into the previous `else` dropping
the additional check.

jakirkham committed:
DOC: Drop `-` from `0`

jakirkham committed:
DOC: Wrap result line to match expected result

cmarmo commented:
The `supervised_learning.rst` file has already been fixed. The modification in `cd_fast.pyx` is not benchmarked. Also the code base has changed a lot.
I'm going to close this pull request. Feel free to open an update one with benchmarks and tests.
